core/chains/evm/gas/fee_history_estimator.go

```diff
@@ -0,0 +1,424 @@
+// blocks. On Ethereum the increase is baseFee * 1.125 per block, however in some chains that may vary.
+func (f *FeeHistoryEstimator) RefreshDynamicPrice() error {
+	ctx, cancel := f.stopCh.CtxCancel(evmclient.ContextWithDefaultTimeout())
+	defer cancel()
+
+	// RewardPercentile will be used for maxPriorityFeePerGas estimations and connectivityPercentile to set the highest threshold for bumping.
+	feeHistory, err := f.client.FeeHistory(ctx, max(f.config.BlockHistorySize, 1), []float64{f.config.RewardPercentile, ConnectivityPercentile})
+	if err != nil {
+		return fmt.Errorf("failed to fetch dynamic prices: %s", err)
+	}
+
+	// eth_feeHistory doesn't return the latest baseFee of the range but rather the latest + 1, because it can be derived from the existing
+	// values. Source: https://github.com/ethereum/go-ethereum/blob/b0f66e34ca2a4ea7ae23475224451c8c9a569826/eth/gasprice/feehistory.go#L235
+	// nextBlock is the latest returned + 1 to be aligned with the base fee value.
+	nextBaseFee := assets.NewWei(feeHistory.BaseFee[len(feeHistory.BaseFee)-1])
+	nextBlock := big.NewInt(0).Add(feeHistory.OldestBlock, big.NewInt(int64(f.config.BlockHistorySize)))
+
+	// If BlockHistorySize is 0 it means priority fees will be ignored from the calculations, so we set them to 0.
+	// If it's not we exclude 0 priced priority fees from the RPC response, even though some networks allow them. For empty blocks, eth_feeHistory
+	// returns priority fees with 0 values so it's safer to discard them in order to pick values from a more representative sample.
+	maxPriorityFeePerGas := assets.NewWeiI(0)
+	priorityFeeThresholdWei := assets.NewWeiI(0)
+	if f.config.BlockHistorySize > 0 {
+		var nonZeroRewardsLen int64 = 0
+		priorityFee := big.NewInt(0)
+		priorityFeeThreshold := big.NewInt(0)
+		for _, reward := range feeHistory.Reward {
+			// reward needs to have values for two percentiles
+			if len(reward) < 2 {
+				return fmt.Errorf("reward size incorrect: %d", len(reward))
+			}
+			// We'll calculate the average of non-zero priority fees
+			if reward[0].Cmp(big.NewInt(0)) > 0 {
+				priorityFee = priorityFee.Add(priorityFee, reward[0])
+				nonZeroRewardsLen += 1
+			}
+			// We take the max value for the bumping threshold
+			if reward[1].Cmp(big.NewInt(0)) > 0 {
+				priorityFeeThreshold = bigmath.Max(priorityFeeThreshold, reward[1])
+			}
+		}
+
+		if nonZeroRewardsLen == 0 || priorityFeeThreshold.Cmp(big.NewInt(0)) == 0 {
+			return nil
+		}
+		priorityFeeThresholdWei = assets.NewWei(priorityFeeThreshold)
+		maxPriorityFeePerGas = assets.NewWei(priorityFee.Div(priorityFee, big.NewInt(nonZeroRewardsLen)))
+	}
+	// baseFeeBufferPercentage is added on top as a safety to catch fluctuations in the next blocks.
+	maxFeePerGas := nextBaseFee.AddPercentage(BaseFeeBufferPercentage).Add(maxPriorityFeePerGas)
+
+	promFeeHistoryEstimatorBaseFee.WithLabelValues(f.chainID.String()).Set(float64(nextBaseFee.Int64()))
```

Comment:
Should we maybe report `nextBaseFee.AddPercentage(BaseFeeBufferPercentage)` here instead? Otherwise the maxFee we report for `promFeeHistoryEstimatorMaxFeePerGas` included the buffered base fee. Otherwise the base fee + priority fee != max fee if someone is trying to use these metrics

Reply:
I wanted to use this metric to track the network's live base fee prices rather than the one we use to construct `maxFeePerGas` since base fee is not used as a primary input for transaction fees. Then we can compare that what we actually sent and measure how effective we are with the estimation.